subsys/serialization/transport/trans_rpmsg.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef TRANS_RPMSG_H_
+#define TRANS_RPMSG_H_
+
+#include <zephyr.h>
+#include <metal/sys.h>
+#include <metal/device.h>
+#include <metal/alloc.h>
+#include <openamp/open_amp.h>
+
+#include "rp_ll.h"
+
+/**
+ * @file
+ * @defgroup rp_transport Remote procedures transport
+ * @{
+ * @brief Remote procedures transport implementation using rpmsg
+ *
+ * API is compatible with rp_trans API. For API documentation
+ * @see rp_trans_tmpl.h
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+struct rp_trans_endpoint {
+	struct rp_ll_endpoint ep;
+	struct k_sem sem;
+	struct k_thread thread;
+	k_thread_stack_t *stack;
+	void *stack_buffer;
+	struct k_fifo fifo;
+	bool running;
+};
+
+typedef void (*rp_trans_receive_handler)(struct rp_trans_endpoint *endpoint,
+	const u8_t *buf, size_t length);
+
+int rp_trans_init(rp_trans_receive_handler callback);
+
+
+void rp_trans_uninit(void);
+
+int rp_trans_endpoint_init(struct rp_trans_endpoint *endpoint,
+	int endpoint_number, size_t stack_size, int prio);
+
+void rp_trans_endpoint_uninit(struct rp_trans_endpoint *endpoint);
+
+#define rp_trans_alloc_tx_buf(endpoint, buf, length) \
+	ARG_UNUSED(endpoint); \
+	u32_t _rp_trans_buf_vla \
+		[(*(length) + sizeof(u32_t) - 1) / sizeof(u32_t)]; \
+	*(buf) = (u8_t *)(&_rp_trans_buf_vla[0])
```

Comment:
Complex logic should not be used in preprocessor macros.

Also, missing while(0) wrapper.

Reply:
This macro is using VLA with `_rp_trans_buf_vla`. Because of that: 1) It cannot be a function - variable will be lost, 2) It cannot be wrapped with `while` - variable will be lost.

Reply:
Could we use malloc instead? Then we won't need the VLA.

Reply:
This doesn't have to be a VLA in any of its current usages, and is generally the same pattern as `NET_BUF_SIMPLE_DEFINE` uses in the Bluetooth stack. It works fine with a static size 99% of the time. The last 1% can use a static maxlen define.

Reply:
`rp_trans_alloc_tx_buf` is used is encoders and it's up the encoder author how much space it needs. Mostly it will be constant, so using `malloc` will be inefficient is this cases. If length is variable then VLA will be used which is faster than `malloc`.

Reply:
This isn't really a problem with net_buf, which has a LOT of encoders working on top of it, solving the same problem. Using VLAs is generally seen as [bad practice](https://www.phoronix.com/scan.php?page=news_item&px=Linux-Kills-The-VLA), and I don't think we should force it upon the encoders. Again, please consider following `NET_BUF_SIMPLE_DEFINE`'s design, it's already proven.

Reply:
But `NET_BUF_SIMPLE_DEFINE` works the same as `rp_trans_alloc_tx_buf`. Only adds some net_buf specific stuff. You can say that `rp_trans_alloc_tx_buf` is stripped version of `NET_BUF_SIMPLE_DEFINE`.